shell/platform/fuchsia/dart_runner/dart_test_component_controller_v2.h

```diff
@@ -0,0 +1,157 @@
+#include <fuchsia/sys/cpp/fidl.h>
+#include <fuchsia/test/cpp/fidl.h>
+#include <lib/async-loop/cpp/loop.h>
+#include <lib/async/cpp/wait.h>
+#include <lib/fdio/namespace.h>
+#include <lib/sys/cpp/component_context.h>
+#include <lib/sys/cpp/service_directory.h>
+#include <lib/zx/timer.h>
+
+#include <lib/fidl/cpp/binding_set.h>
+#include "lib/fidl/cpp/binding.h"
+#include "runtime/dart/utils/mapped_resource.h"
+#include "third_party/dart/runtime/include/dart_api.h"
+
+namespace dart_runner {
+
+/// Starts a Dart component written in CFv2.
+class DartTestComponentControllerV2
+    : public fuchsia::component::runner::ComponentController,
+      public fuchsia::test::Suite {
+  using DoneCallback = fit::function<void(DartTestComponentControllerV2*)>;
+
+ public:
+  DartTestComponentControllerV2(
+      fuchsia::component::runner::ComponentStartInfo start_info,
+      std::shared_ptr<sys::ServiceDirectory> runner_incoming_services,
+      fidl::InterfaceRequest<fuchsia::component::runner::ComponentController>
+          controller,
+      DoneCallback done_callback);
+
+  ~DartTestComponentControllerV2() override;
+
+  /// Sets up the controller.
+  ///
+  /// This should be called before |Run|.
+  void SetUp();
+
+  /// |Suite| protocol implementation.
+  void GetTests(
+      fidl::InterfaceRequest<fuchsia::test::CaseIterator> iterator) override;
+
+  /// |Suite| protocol implementation.
+  void Run(std::vector<fuchsia::test::Invocation> tests,
+           fuchsia::test::RunOptions options,
+           fidl::InterfaceHandle<fuchsia::test::RunListener> listener) override;
+
+  fidl::InterfaceRequestHandler<fuchsia::test::Suite> GetHandler() {
+    return suite_bindings_.GetHandler(this, loop_->dispatcher());
+  }
+
+ private:
+  /// Helper for actually running the Dart main. Returns true if successful,
+  /// false otherwise.
+  bool RunDartMain();
+
+  /// Creates and binds the namespace for this component. Returns true if
+  /// successful, false otherwise.
+  bool CreateAndBindNamespace();
+
+  bool SetUpFromKernel();
+  bool SetUpFromAppSnapshot();
+
+  bool CreateIsolate(const uint8_t* isolate_snapshot_data,
+                     const uint8_t* isolate_snapshot_instructions);
+
+  // |ComponentController|
+  void Kill() override;
+  void Stop() override;
+
+  // Idle notification.
+  void MessageEpilogue(Dart_Handle result);
+  void OnIdleTimer(async_dispatcher_t* dispatcher,
+                   async::WaitBase* wait,
+                   zx_status_t status,
+                   const zx_packet_signal* signal);
+
+  // |CaseIterator|
+  class CaseIterator final : public fuchsia::test::CaseIterator {
+   public:
+    CaseIterator(fidl::InterfaceRequest<fuchsia::test::CaseIterator> request,
+                 async_dispatcher_t* dispatcher,
+                 fit::function<void(CaseIterator*)> done_callback);
+
+    void GetNext(GetNextCallback callback) override;
+
+   private:
+    bool first_case = true;
```

Comment:
Sorry, I screwed up in my comment, it should be first_case_ with a trailing underscore since it's a private field.